### PR TITLE
feat(medusa): cache modules

### PR DESCRIPTION
**Pull request by fPolic**
**What**
- add in-memory cache module
- add Redis cache module
- remove the cache service from the core
- set in-memory cache module as default?
- use cache module in `medusa-source-shopify`
- use cache module in `medusa-plugin-contentful`

--- 

RESOLVES CORE-1088
RESOLVES CORE-1087
RESOLVES CORE-1171
RESOLVES CORE-1172
RESOLVES CORE-1224

### CODE REVIEW

File: packages/cache-inmemory/.npmignore
Comment (adrien2p) [inline packages/cache-inmemory/.npmignore:1]: todo: rm in favor of the package.json comment

File: packages/cache-inmemory/package.json
Comment (adrien2p) [inline packages/cache-inmemory/package.json:11]: todo: add files: ["dist"]

File: packages/cache-inmemory/package.json
Comment (adrien2p) [inline packages/cache-inmemory/package.json:35]: todo:I would keep this as a peer deps like the other packages. We did the same in the typeorm pr for the other modules

File: packages/cache-inmemory/src/services/__tests__/inmemory-cache.js
Comment (adrien2p) [inline packages/cache-inmemory/src/services/__tests__/inmemory-cache.js:57]: suggestion: should we extract the sleep util from the core to go into medusa-core-utils? and use it where we needs?

File: packages/cache-inmemory/src/services/__tests__/inmemory-cache.js
Comment (adrien2p) [inline packages/cache-inmemory/src/services/__tests__/inmemory-cache.js:62]: question: It seams to be the same as the previous one since none of them passes a TTL no?

File: packages/cache-inmemory/src/services/__tests__/inmemory-cache.js
Comment (adrien2p) [inline packages/cache-inmemory/src/services/__tests__/inmemory-cache.js:76]: suggestion: 1 should be enought to not run the test for too long

File: packages/cache-inmemory/src/services/__tests__/inmemory-cache.js
Comment (adrien2p) [inline packages/cache-inmemory/src/services/__tests__/inmemory-cache.js:83]: comment: change depending on the previous comment

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (adrien2p) [inline packages/cache-inmemory/src/services/inmemory-cache.ts:20]: question: Not sure it is necessary to extends this class and hold the manager in the class. 
If some provide a db cache, then he will have a local transaction and if the core fail the result will still be cached which I think still makes sense if it succeed to compute something. But overall I am not sure we need that

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (adrien2p) [inline packages/cache-inmemory/src/services/inmemory-cache.ts:24]: comment: rm depending on previous acceptation

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (adrien2p) [inline packages/cache-inmemory/src/services/inmemory-cache.ts:28]: todo: use Map instead for dynamic keys it is better and more performant, specially for a lot of values

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (adrien2p) [inline packages/cache-inmemory/src/services/inmemory-cache.ts:37]: comment: rm depending on acceptation of previous comment

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (adrien2p) [inline packages/cache-inmemory/src/services/inmemory-cache.ts:56]: suggestion: store the now at the beginning of the method just for safety if it is close to the expire :D wdyt?

File: packages/cache-redis/.npmignore
Comment (adrien2p) [inline packages/cache-redis/.npmignore:1]: todo: rm for the same reason as the previous module

File: packages/cache-redis/package.json
Comment (adrien2p) [inline packages/cache-redis/package.json:11]: todo: files: ["dist"]

File: packages/cache-redis/package.json
Comment (adrien2p) [inline packages/cache-redis/package.json:35]: todo: same as previous moodule

File: packages/cache-redis/src/services/redis-cache.ts
Comment (adrien2p) [inline packages/cache-redis/src/services/redis-cache.ts:20]: suggestion: Same question as the previous module

File: packages/cache-redis/src/services/redis-cache.ts
Comment (adrien2p) [inline packages/cache-redis/src/services/redis-cache.ts:93]: todo: rm

File: packages/medusa/src/loaders/module.ts
Comment (adrien2p) [inline packages/medusa/src/loaders/module.ts:24]: 😂 praise

**Review by adrien2p — CHANGES_REQUESTED** [review]

File: packages/cache-inmemory/.npmignore
Comment (adrien2p) [inline packages/cache-inmemory/.npmignore]: todo: I mean you can remove the entire file 😄

File: packages/cache-inmemory/package.json
Comment (adrien2p) [inline packages/cache-inmemory/package.json:32]: todo: missing `^`

**Review by adrien2p — CHANGES_REQUESTED** [review]

File: packages/cache-inmemory/src/services/__tests__/inmemory-cache.js
Comment (fPolic) [reply packages/cache-inmemory/src/services/__tests__/inmemory-cache.js:62]: in the previous case TTL is passed on L:54

File: packages/modules-sdk/src/loaders/module-loader.ts
Comment (carlos-r-l-rodrigues) [inline packages/modules-sdk/src/loaders/module-loader.ts:25]: I'm correcting this typo in another PR and this file is changing quite a bit, so I believe it is better to revert these changes to avoid conflict.

File: packages/cache-inmemory/package.json
Comment (carlos-r-l-rodrigues) [inline packages/cache-inmemory/package.json:22]:
```suggestion
    "jest": "^25.5.4",
```

File: packages/cache-inmemory/src/index.ts
Comment (carlos-r-l-rodrigues) [inline packages/cache-inmemory/src/index.ts:9]: FYI this [PR](https://github.com/medusajs/medusa/pull/3329/files#diff-c7b62e8108f9340c97e8576ca2692105f91742dd659d457f903ce02cb036ca25R89) will change that export to accept multiple services. Also the type ModuleExports is located at "@medusajs/module-sdk".
No action needed for now. If this is merged before I can apply the changes.

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (carlos-r-l-rodrigues) [inline packages/cache-inmemory/src/services/inmemory-cache.ts:72]: set it to `this.timoutRefs` is missing?

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (carlos-r-l-rodrigues) [inline packages/cache-inmemory/src/services/inmemory-cache.ts:93]: it should loop and call clearTimeout for all references

File: packages/cache-redis/src/services/redis-cache.ts
Comment (carlos-r-l-rodrigues) [inline packages/cache-redis/src/services/redis-cache.ts:29]: I think we can also support "isolated" resources here.
I can apply the changes after the PR is merged if you prefer.
[Here](https://github.com/medusajs/medusa/pull/3329/files#diff-d2d3c280dba3a849f2a36c762afcef78f9a556cf3c94b3d261b6425898bb5ac8) is an example of the loader creating and registering it's own connection.

File: packages/cache-redis/src/services/redis-cache.ts
Comment (carlos-r-l-rodrigues) [inline packages/cache-redis/src/services/redis-cache.ts:69]: should we also support wildcards "*" on the in-memory cache?

File: packages/modules-sdk/src/loaders/module-loader.ts
Comment (fPolic) [reply packages/modules-sdk/src/loaders/module-loader.ts]: reverted in favour of that PR

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (fPolic) [reply packages/cache-inmemory/src/services/inmemory-cache.ts:72]: Good catch, fixed in https://github.com/medusajs/medusa/pull/3187/commits/2284503bc5dc175f2aea36ce23362b41e4276d10

File: packages/cache-redis/src/services/redis-cache.ts
Comment (olivermrbl) [reply packages/cache-redis/src/services/redis-cache.ts:29]: In a somewhat similar context, should we remove the dependency on the `redisClient` from the core and use the same approach as in the Redis event bus i.e. create a dedicated connection in the module loader?

You can check it out in [this commit](https://github.com/medusajs/medusa/pull/2599/commits/1753cc20c266eacb2926079281c7b22d14b5b50a). 

@fPolic what do you think?

File: packages/cache-redis/src/services/redis-cache.ts
Comment (olivermrbl) [inline packages/cache-redis/src/services/redis-cache.ts:27]: thought(non-blocking): Should we expose a `namespace` option to group your Medusa cache entries? We could default to `medusa::` or something similar.

File: packages/cache-inmemory/src/types/index.ts
Comment (olivermrbl) [inline packages/cache-inmemory/src/types/index.ts:16]: suggestion(non-blocking): Rename to `ttl`. The default one comes from within the class.

**Review by olivermrbl — COMMENTED** [review]
Great work @fPolic! Have added a couple of comments, that I'd like you to address :)

**Review by olivermrbl — CHANGES_REQUESTED** [review]
And a few chores, that I missed in first review:
- READMEs for both modules
- changeset

💪

File: packages/cache-redis/src/services/redis-cache.ts
Comment (fPolic) [reply packages/cache-redis/src/services/redis-cache.ts]: I've added the namespace option with the default `medusa:` namespace as you suggested but I think a default namespace could be potentially easy to miss (e.g. user searches for `my-key*` instead of `medusa:my-key*`) so maybe just default to an empty string, wdyt?

File: packages/cache-redis/src/services/redis-cache.ts
Comment (fPolic) [reply packages/cache-redis/src/services/redis-cache.ts]: yes, this makes more sense 👍 
I've added the loader in https://github.com/medusajs/medusa/pull/3187/commits/2a092e5b407687f41487901bf5b33b90b69a8861

File: packages/cache-redis/src/services/redis-cache.ts
Comment (fPolic) [reply packages/cache-redis/src/services/redis-cache.ts:70]: I think it's a good idea, but is that something we should "enforce" with the `ICacheInterface` somehow to ensure other caching modules implement that behaviour as well, wdyt?

File: packages/cache-redis/src/services/redis-cache.ts
Comment (olivermrbl) [reply packages/cache-redis/src/services/redis-cache.ts]: That's true. But if we're explicit about it in the README, we should be fine no?

Comment (fPolic) [reply packages/cache-redis/src/services/redis-cache.ts]: Sounds good, I added it to the Readme

**Comment by olivermrbl** [comment]
Tested both modules locally, and the setup and functionality are working as expected. Great work @fPolic!

@carlos-r-l-rodrigues can I get you to give this another review when time allows for it? You've previously added a comment on the PR.

File: packages/cache-redis/src/services/redis-cache.ts
Comment (olivermrbl) [reply packages/cache-redis/src/services/redis-cache.ts:65]: Do you think we should hold off on this for now? We will recommend the Redis cache for production environments - even though the in-memory module might be useful for some cases. And then be very explicit about this in our documentation of the in-memory module.

I don't think we should let complex regex acrobatics stand in the way of merging this soon. What's your take?

**Comment by olivermrbl** [comment]
@fPolic Can I get you to investigate the failing pipeline? Not the CLI one.

File: packages/cache-redis/src/services/redis-cache.ts
Comment (fPolic) [reply packages/cache-redis/src/services/redis-cache.ts:65]: @olivermrbl sounds good! I will add a ticket to implement this in a follow-up iteration and add the recommendation to the Readme.

**Comment by adrien2p** [comment]
> @adrien2p we also need you to give this a second review. No rush, but would be great to merge later today, if possible :)

Yes sure, i ll re review it now

File: packages/cache-redis/src/services/redis-cache.ts
Comment (adrien2p) [reply packages/cache-redis/src/services/redis-cache.ts:65]: I think that it is a very good idea to include in another iteration 👍

**Review by adrien2p — CHANGES_REQUESTED** [review]
Nice work man, few suggestions and todo

File: packages/cache-inmemory/src/index.ts
Comment (carlos-r-l-rodrigues) [reply packages/cache-inmemory/src/index.ts:9]: Will remove that @adrien2p , indeed only services are required to be exported

File: integration-tests/api/medusa-config.js
Comment (fPolic) [reply integration-tests/api/medusa-config.js:11]: think it's still needed for subscribers to work in the test env

File: packages/cache-inmemory/src/services/__tests__/inmemory-cache.js
Comment (fPolic) [reply packages/cache-inmemory/src/services/__tests__/inmemory-cache.js:86]: Fixed the previous one, now they are different

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (fPolic) [reply packages/cache-inmemory/src/services/inmemory-cache.ts]: I had the check but removed it due to: https://github.com/medusajs/medusa/pull/3187#discussion_r1116855318

File: packages/cache-inmemory/src/services/__tests__/inmemory-cache.js
Comment (fPolic) [reply packages/cache-inmemory/src/services/__tests__/inmemory-cache.js:70]: I don't think this oneliner is something for `medusa-core-utils` especially since you would use this in tests mostly and it's not domain related, but if you think we should extract it there we can.

File: packages/cache-redis/src/services/redis-cache.ts
Comment (fPolic) [reply packages/cache-redis/src/services/redis-cache.ts:65]: @olivermrbl implemented `*` matching for deletion today so now we have the same functionality in redis and inmemory cache modules

File: packages/cache-redis/src/loaders/index.ts
Comment (fPolic) [reply packages/cache-redis/src/loaders/index.ts]: Prefixing and identifier sound good 👍 , I renamed the key to `cacheRedisConnection` for now

File: packages/cache-redis/src/services/redis-cache.ts
Comment (olivermrbl) [reply packages/cache-redis/src/services/redis-cache.ts:65]: @fPolic That's great!

File: packages/cache-inmemory/src/services/__tests__/inmemory-cache.js
Comment (adrien2p) [reply packages/cache-inmemory/src/services/__tests__/inmemory-cache.js:70]: Don’t have strong opinions so i let you choose

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (adrien2p) [reply packages/cache-inmemory/src/services/inmemory-cache.ts]: Got it make sense 💪

File: packages/cache-inmemory/package.json
Comment (olivermrbl) [inline packages/cache-inmemory/package.json:16]: question: `publishConfig` and `files` are typically added as two separate props at the root level. Can you verify that this works as expected? :)

File: packages/modules-sdk/src/definitions.ts
Comment (olivermrbl) [inline packages/modules-sdk/src/definitions.ts]: comment: Nice catch 😄

**Review by olivermrbl — APPROVED** [review]
LGTM, great work - pending @adrien2p's approval 🥳

File: packages/medusa-source-shopify/src/services/shopify-product.js
Comment (fPolic) [reply packages/medusa-source-shopify/src/services/shopify-product.js:114]: Good catch, I think exposing clients would kinda break the interface because some clients don't have that functionality and some modules don't have clients at all. But in this case, I would suggest we reimplement this with a simple Map on service and keep the cache interface clean so it could be implemented across different technologies, wdyt?

Comment (fPolic) [reply packages/medusa-source-shopify/src/services/shopify-product.js:114]: `this.cacheService` is an instance of `ShopifyCacheService` which is using a cache module as an underlying store but have this methods

File: packages/cache-inmemory/jest.config.js
Comment (adrien2p) [inline packages/cache-inmemory/jest.config.js:4]: question: Does this works? Since this config exclude all tests files? Should we have a separate tsconfig.spec.json as we do in the plugins?

File: packages/medusa-source-shopify/src/services/__tests__/shopify-product.js
Comment (adrien2p) [inline packages/medusa-source-shopify/src/services/__tests__/shopify-product.js:49]: todo: We still have those methods that does not exists on the cache service

**Review by adrien2p — CHANGES_REQUESTED** [review]

File: packages/cache-inmemory/src/services/inmemory-cache.ts
Comment (olivermrbl) [reply packages/cache-inmemory/src/services/inmemory-cache.ts:1]: Yes, I agree, this is a reasonable approach. And should be doable before v1.8.0 release. 

So the first iteration would be to up the `@medusajs/types` package and centralize all the imports from `@medusajs/medusa`, which our modules are currently relying on. And follow up iterations could involve moving other common types, but I wouldn't worry about that now.

Comment (carlos-r-l-rodrigues) [reply packages/cache-inmemory/src/services/inmemory-cache.ts:1]: I'm not sure if I followed the "import something from index" statement. The module repository won't import the database from the core. But moving types the modules need to a different package is something we'll need, and apparently it isn't a small task.

Comment (adrien2p) [reply packages/cache-inmemory/src/services/inmemory-cache.ts:1]: when you import something from the index of the core, every import in the index and related files gets evaluated as well. The idea here, is not to move everything yet, but just whats needed, so for that pr only the cache service interface that is at the moment imported from the core

File: packages/cache-inmemory/jest.config.js
Comment (adrien2p) [reply packages/cache-inmemory/jest.config.js:4]: If it works, let's keep it that way, I would have just thought that since they are excluded in the tsconfig it might be a problem 😂 Let's keep it that way and see later to harmonised the different plugins and modules as they have or not a tsconfig.spec depending.

**Review by adrien2p — APPROVED** [review]

**Pull request merged**
